Source/DafnyCore/Resolver/ModuleResolver.cs

```diff
@@ -1583,6 +1571,86 @@ public partial class ModuleResolver {
       }
     }
 
+    private void CheckForCyclesAmongRedirectingTypes(RedirectingTypeDecl dd, HashSet<ICallable> cycleErrorHasBeenReported) {
+      var enclosingModule = dd.EnclosingModule;
+      if (enclosingModule.CallGraph.GetSCCSize(dd) != 1) {
+        var r = enclosingModule.CallGraph.GetSCCRepresentative(dd);
+        if (cycleErrorHasBeenReported.Contains(r)) {
+          // An error has already been reported for this cycle, so don't report another.
+          // Note, the representative, "r", may itself not be a const.
+        } else if (dd is NewtypeDecl || dd is SubsetTypeDecl) {
+          ReportCallGraphCycleError(dd, $"recursive constraint dependency involving a {dd.WhatKind}");
+          cycleErrorHasBeenReported.Add(r);
+        }
+      }
+    }
+
+    /// <summary>
+    /// Return "declarations" in sorted order, so that each declaration is preceded by those on which it depends.
+    /// It is assumed that "declarations" are all from the same module and that the call graph of that module has been constructed.
+    /// If there's some given declaration that's not in the call graph, then that's fine; it just means that the call
+    /// graph does not constrain the output order of that declaration.
+    /// </summary>
+    private List<TopLevelDecl> TopologicallySortedTopLevelDecls(List<TopLevelDecl> declarations) {
+      // Dependency information among the declarations is stored in the enclosing module's call graph. To get to that
+      // call graph, we need to have the module declaration, which we obtain from one of the given declarations:
+      if (declarations.Count == 0) {
+        return declarations;
+      }
+      var enclosingModule = declarations[0].EnclosingModuleDefinition;
+      Contract.Assert(declarations.TrueForAll(decl => decl.EnclosingModuleDefinition == enclosingModule));
+
+      // From the module declaration, we get the components sorted according to how they depend on each other.
+      var sortedComponents = enclosingModule.CallGraph.TopologicallySortedComponents();
+
+      // But... this is a list of ICallable's (which also includes MemberDecl's and may not contain all TopLevelDecl's).
```

Comment:
Why would the call-graph not contain all ICallables that are in the module?

Reply:
> Why would the call-graph not contain all ICallables that are in the module?

For some reason, the system module is different. Only some of its declarations go through the regular resolution pipeline. `object` and `array` are missing.

This method was written to be general. But you're right that, here, we need it only for the declarations of subset types and newtypes. So, we could specialize this method and simplify its implementation.

One awkward thing is that `ResolveTopLevelDecls_Core` only gets a list of declarations, not the enclosing module. So, it would still be nice to have `TopologicallySortedTopLevelDecls` as a separate method, so that looking into `declarations[0].ModuleDefinition` would be done there.

Reply:
> One awkward thing is that ResolveTopLevelDecls_Core only gets a list of declarations, not the enclosing module.

Looking at the code we can pass the module as well. And you could replace `CallGraphBuilder.Build(declarations, reporter);` with `CallGraphBuilder.Build(ModuleDefinition.AllDeclarationsAndNonNullTypeDecls(module.TopLevelDecls), reporter);` so it's fully filled for the system module as well.

And then `declarations = TopologicallySortedTopLevelDecls(declarations);` would be `var declarations = module.CallGraph..TopologicallySortedComponents().Where(decl => decl is SubsetTypeDecl or NewtypeDecl)`